gufe/custom_json.py

```diff
@@ -0,0 +1,162 @@
+# This file was originally part of OpenPathSampling/SimStore, which is
+# distributed under the MIT license.
+# Portions Copyright (c) 2014-2022 the contributors to OpenPathSampling
+# Permissions are the same as those listed in the gufe LICENSE
+
+import json
+import functools
+
+
+class JSONSerializerDeserializer(object):
+    """
+    Tools to serialize and deserialize objects as JSON.
+
+    This wrapper object is necessary so that we can register new codecs
+    after the original initialization.
+
+    Attributes
+    ----------
+    encoder:
+        subclass of ``JSONEncoder``; use as ``json.dumps(obj, cls=encoder)``
+    decoder:
+        subclass of ``JSONDecoder``; use as ``json.loads(string,
+        cls=decoder)``
+
+    Parameters
+    ----------
+    codecs : list of :class:`.JSONCodec`s
+        codecs supported
+    """
+    def __init__(self, codecs):
+        self.codecs = []
+        for codec in codecs:
+            self.add_codec(codec)
+
+        self.encoder, self.decoder = self._set_serialization()
+
+    def _set_serialization(self):
+        encoder, decoder = custom_json_factory(self.codecs)
+        self._serializer = functools.partial(json.dumps, cls=encoder)
+        self._deserializer = functools.partial(json.loads, cls=decoder)
+        return encoder, decoder
+
+    def add_codec(self, codec):
+        """Add a new codec to the supported codecs
+
+        Parameters
+        ----------
+        codec : :class:`.JSONCodec`
+            codec to add
+        """
+        if codec in self.codecs:
+            return
+
+        if codec is not None:
+            self.codecs.append(codec)
+
+        self.encoder, self.decoder = self._set_serialization()
+
+
+    def serializer(self, obj):
+        """Callable that dumps to JSON"""
+        return self._serializer(obj)
+
+    def deserializer(self, string):
+        """Callable to loads JSON"""
+        return self._deserializer(string)
+
+
+def custom_json_factory(coding_methods):
```

Comment:
I'd be tempted to make this private and keep the API surface smaller (or document and test it properly)

Reply:
It's tested in anything that inherits from `CustomJSONTest`, in the `test_round_trip` method.

If, by documentation, you mean on the dynamically-defined classes, I disagree. I don't want to override the superclass docs, which you get now if you do

```python
help(JSON_HANDLER.encoder)
```

The only possible reason to override the super's docs is that in decoder, technically you'll get an error if you specify `object_hook` (since we're already specifying it).

Reply:
Re docs: This function isn't well documented, it took me a while to get what was going on.

Reply:
Do you mean that it isn't clear that it is a factory function returning dynamically created classes? Or do you mean that mechanism of the `default` / `object_hook` methods are unclear?

Reply:
Added more docs on this method in [e320f93](https://github.com/OpenFreeEnergy/gufe/pull/68/commits/e320f933332f0732af998cbe71f0aac567affaab). Honestly not sure what else to do if this doesn't explain it well enough.